### ISSUE TITLE
123

### ISSUE THREAD

**Issue by SteveHong1901** [SteveHong1901/STAT0023-Coursework-1#2]
# Import the necessary module from Huggingface's transformers library
from transformers import AutoTokenizer

# Define a custom Dataset class extending PyTorch's Dataset class
class Dataset(torch.utils.data.Dataset):
    def __init__(self, X, y, max_length, id_to_label):
        # Initialize the tokenizer with a pretrained DistilBERT model
        self.tokenizer = AutoTokenizer.from_pretrained("distilbert-base-uncased")
        
        # Store the input data and labels
        self.x = X
        self.y = y
        
        # Store the dictionary mapping id to label
        self.id_to_label = id_to_label

    def __getitem__(self, idx):
        # Returns a tokenized and aligned version of the item at the given index
        return self.tokenize_and_align_labels(self.x[idx], self.y[idx])
    
    def __len__(self):
        # Returns the length of the input data
        return len(self.x)
    
    def tokenize_and_align_labels(self, x_el, y_el):
        # Tokenize input data with truncation and padding, also split into words
        tokenized_inputs = self.tokenizer(x_el, truncation=True, padding="max_length", is_split_into_words=True, max_length=max_length)
        
        # Convert to tensor format
        data = {key: torch.tensor(val) for key, val in tokenized_inputs.items()}
        
        # Store the labels
        ner_tags = y_el
        
        # Initialize lists to store label IDs
        labels = []
        label_ids = []
        
        # Map tokens to their respective word
        word_ids = tokenized_inputs.word_ids(batch_index=0)
        
        # Variable to keep track of the previous word's index
        previous_word_idx = None
        
        for word_idx in word_ids:
            # Set the special tokens to -100
            if word_idx is None:
                label_ids.append(-100)
            elif word_idx != previous_word_idx:  # Only label the first token of a given word
                label_ids.append(ner_tags[word_idx])
            else:
                label_ids.append(-100)  # If it's not the first token of a word, set label to -100
            previous_word_idx = word_idx
            
        labels.append(label_ids)
        
        # Store the labels as a tensor
        data["labels"] = torch.tensor(labels).squeeze()
        
        return data

# Create a dataset instance with given input data, labels, maximum length, and id-to-label dictionary
total_dataset = Dataset(X, y, max_length, id_to_label)
